### PR TITLE
Add custom annotations support

### PR DESCRIPTION
**Pull request by kxfang**
Allows adding custom annotations and associated
prefixes to reduce repetition.

Addresses #144

### CODE REVIEW

**Comment by kxfang** [comment]
Hmm, noticed a couple of issues. Hang on.

**Comment by kxfang** [comment]
Should be good to go. Note that I tried annotating @DeepLink with @DeepLinkSpec so that we don't need to handle it as a special case in the processor, but for whatever reason it just refuses to get picked up by `roundEnv.getElementsAnnotatedWith(DEEP_LINK_SPEC_CLASS)`

**Comment by kxfang** [comment]
Hmm, not sure how my latest change managed to break the build. Can't reproduce it locally

File: sample/src/main/java/com/airbnb/deeplinkdispatch/sample/CustomPrefixesActivity.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                showToast("class id== " + idString);
+            } else {
+                showToast("no id in the deeplink");
+            }
+        } else {
+            showToast("no deep link :( ");
+        }
+    }
+
+
```
Comment (felipecsl) [inline sample/src/main/java/com/airbnb/deeplinkdispatch/sample/CustomPrefixesActivity.java]: nit: double line break

File: deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java [hunk 2]
```diff
@@ -65,6 +65,88 @@
...
+              + "      if (entry.matches(uri)) {\n"
+              + "        return entry;\n"
+              + "      }\n"
+              + "    }\n"
+              + "    return null;\n"
+              + "  }\n"
+              + "}"));
+}
+
+
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java]: nit: double line break

**Comment by felipecsl** [comment]
Just triggered another build. It might have been Travis issues

File: sample/src/main/java/com/airbnb/deeplinkdispatch/sample/CustomPrefixesActivity.java (hunk 1 above)
Comment (kxfang) [inline sample/src/main/java/com/airbnb/deeplinkdispatch/sample/CustomPrefixesActivity.java]: Done

File: deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java (hunk 2 above)
Comment (kxfang) [inline deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkProcessorTest.java]: Done

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 3]
```diff
@@ -98,24 +107,55 @@
...
+    Set<Element> customAnnotatedElements = new HashSet<>();
+    for (Element customAnnotation : roundEnv.getElementsAnnotatedWith(DEEP_LINK_SPEC_CLASS)) {
+      ElementKind kind = customAnnotation.getKind();
+      if (kind != ElementKind.ANNOTATION_TYPE) {
+        error(customAnnotation, "Only annotation types can be annotated with @%s",
+                DEEP_LINK_SPEC_CLASS.getSimpleName());
+      }
+      String[] prefix = customAnnotation.getAnnotation(DEEP_LINK_SPEC_CLASS).prefix();
+      if (Utils.hasEmptyOrNullString(prefix)) {
+        error(customAnnotation, "Prefix property cannot have null or empty strings");
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: need a unit test for this case

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 4]
```diff
@@ -98,24 +107,55 @@
...
+      }
+      if (prefix.length == 0) {
+        error(customAnnotation, "Prefix property cannot be empty");
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: and this

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java (hunk 3 above)
Comment (kxfang) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: Done

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java (hunk 4 above)
Comment (kxfang) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: Done

File: checkstyle.xml [hunk 5]
```diff
@@ -14,6 +14,12 @@
         <property name="message" value="Line has trailing spaces."/>
     </module>
 
+    <!-- Checks for consecutive blank lines -->
+    <module name="RegexpMultiline">
+        <property name="format" value="^[ \t\r]*\n[ \t\r]*$"/>
+        <property name="message" value="Consecutive blank lines."/>
```
Comment (felipecsl) [inline checkstyle.xml]: 👍

**Comment by felipecsl** [comment]
Thanks Kevin and Mike Evans. I'll release this as 3.0.0-rc2

**Comment by kxfang** [comment]
Thanks! When should I look for the release?

**Comment by felipecsl** [comment]
Just pushes to central. Should be live in a couple hours

**Comment by kxfang** [comment]
Perfect, thanks!

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Utils.java [hunk 6]
```diff
@@ -15,4 +15,13 @@ static String decapitalize(String str) {
       return str;
     }
   }
+
+  static boolean hasEmptyOrNullString(String[] strings) {
+    for (String s : strings) {
+      if (s == null || s.isEmpty()) {
```
Comment (OleksandrKucherenko) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Utils.java]: TextUtils.isEmpty()

Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Utils.java]: `TextUtils` is part of the android sdk. This is a pure java library, so we don't have access to it

**Pull request merged**